regression/cuda/test_conv2d.c

```diff
@@ -0,0 +1,283 @@
+// Copyright (c) 2019, University of Washington All rights reserved.
+// 
+// Redistribution and use in source and binary forms, with or without modification,
+// are permitted provided that the following conditions are met:
+// 
+// Redistributions of source code must retain the above copyright notice, this list
+// of conditions and the following disclaimer.
+// 
+// Redistributions in binary form must reproduce the above copyright notice, this
+// list of conditions and the following disclaimer in the documentation and/or
+// other materials provided with the distribution.
+// 
+// Neither the name of the copyright holder nor the names of its contributors may
+// be used to endorse or promote products derived from this software without
+// specific prior written permission.
+// 
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+// ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+// WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+// DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR
+// ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+// (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+// LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+// ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+#include "test_conv2d.h"
+#include <math.h>
+
+#define ALLOC_NAME "default_allocator"
+
+#define MIN_N 4
+#define MAX_N 1024
+
+#define MIN_F 1
+#define MAX_F 128
+
+#define MIN_P 0
+#define MAX_P 8
+
+#define MIN_S 1
+#define MAX_S 8
+
+#define MAX_VALUE_MAGNITUDE 1e6
+
+int generate_inclusive(int min, int max)
+{
+        int range = (max - min + 1);
+        return (rand() % range) + min;
+}
+
```

Comment:
You should see how I've started generating random numbers here: https://github.com/bespoke-silicon-group/bsg_f1/blob/9cdfa2862ca547dc6d9b5fc817db1bf66ba70403/regression/cuda/test_vector_matrix_mul.cpp#L95

It's worth switching to C++ to get nice things :p

Reply:
Hmmm I'm not convinced C++ would be very much nicer in this particular case - since the limits are different each time I generate it, I'd have to make a new generator object every time, and then I'd have to have a random seed each time.